geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java

```diff
@@ -2952,6 +2870,102 @@ public void txApplyPut(Operation p_putOp, Object key, Object nv, boolean didDest
+      boolean cbEventInPending, boolean opCompleted) {
+    // At this point we should only apply the update if the entry exists
+    RegionEntry re = getEntry(key); // Fix for bug 32347.
+    if (re != null) {
+      synchronized (re) {
+        if (!re.isRemoved()) {
+          opCompleted = true;
+          putOp = putOp.getCorrespondingUpdateOp();
+          // Net writers are not called for received transaction data
+          final int oldSize = owner.calculateRegionEntryValueSize(re);
+          if (cbEvent != null) {
+            cbEvent.setRegionEntry(re);
+            cbEvent.setOldValue(re.getValueInVM(owner)); // OFFHEAP eei
+          }
+
+          boolean clearOccured = false;
+          // Set RegionEntry updateInProgress
+          if (owner.indexMaintenanceSynchronous) {
+            re.setUpdateInProgress(true);
+          }
+          try {
+            txRemoveOldIndexEntry(putOp, re);
+            if (didDestroy) {
+              re.txDidDestroy(owner.cacheTimeMillis());
+            }
+            if (txEvent != null) {
+              txEvent.addPut(putOp, owner, re, re.getKey(), newValue, aCallbackArgument);
+            }
+            re.setValueResultOfSearch(putOp.isNetSearch());
+            try {
+              processAndGenerateTXVersionTag(owner, cbEvent, re, txEntryState);
+              {
+                re.setValue(owner,
+                    re.prepareValueForCache(owner, newValue, cbEvent, !putOp.isCreate()));
+              }
+              if (putOp.isCreate()) {
+                owner.updateSizeOnCreate(key, owner.calculateRegionEntryValueSize(re));
+              } else if (putOp.isUpdate()) {
+                // Rahul : fix for 41694. Negative bucket size can also be
+                // an issue with normal GFE Delta and will have to be fixed
+                // in a similar manner and may be this fix the the one for
+                // other delta can be combined.
+                {
+                  owner.updateSizeOnPut(key, oldSize, owner.calculateRegionEntryValueSize(re));
+                }
+              }
+            } catch (RegionClearedException rce) {
+              clearOccured = true;
+            }
+            {
+              long lastMod = owner.cacheTimeMillis();
+              EntryLogger.logTXPut(_getOwnerObject(), key, nv);
+              re.updateStatsForPut(lastMod, lastMod);
+              owner.txApplyPutPart2(re, re.getKey(), lastMod, false, didDestroy, clearOccured);
+            }
+          } finally {
+            if (re != null && owner.indexMaintenanceSynchronous) {
+              re.setUpdateInProgress(false);
+            }
+          }
+          if (!clearOccured) {
+            lruEntryUpdate(re);
+          }
+        }
+      }
+      if (didDestroy && !opCompleted) {
+        owner.txApplyInvalidatePart2(re, re.getKey(), true, false /* clear */);
+      }
+    }
+    if (invokeCallbacks) {
```

Comment:
This will invoke CacheListeners too...Does this have any  implications?

Reply:
You are right. Fix is revised to only send event to client if operation is not applied to the cache.